src/ipc/ipc4/helper.c

```diff
@@ -453,32 +459,85 @@ int ipc_comp_connect(struct ipc *ipc, ipc_pipe_comp_connect *_connect)
 	if (!cpu_is_me(source->ipc_config.core) && !cross_core_bind)
 		return ipc4_process_on_core(source->ipc_config.core, false);
 
-	/* these might call comp_ipc4_get_attribute_remote() if necessary */
-	ret = comp_get_attribute(source, COMP_ATTR_BASE_CONFIG, &source_src_cfg);
-	if (ret < 0) {
-		tr_err(&ipc_tr, "failed to get base config for module %#x", dev_comp_id(source));
-		return IPC4_FAILURE;
+	size = MODULE_MAX_SINKS * sizeof(struct ipc4_output_pin_format) +
+		MODULE_MAX_SOURCES * sizeof(struct ipc4_input_pin_format);
+
+	/* get obs from the base config extension if the src queue ID is non-zero */
+	if (bu->extension.r.src_queue) {
+		struct ipc4_output_pin_format out_fmt;
+		size_t output_fmt_offset;
+
+		src_basecfg_ext = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+					  sizeof(*src_basecfg_ext) + size);
+		if (!src_basecfg_ext)
+			return IPC4_OUT_OF_MEMORY;
+
+		ret = comp_get_attribute(source, COMP_ATTR_BASE_CONFIG_EXT, src_basecfg_ext);
+		if (!ret) {
+			size_t in_fmt_size = sizeof(struct ipc4_input_pin_format);
+			size_t out_fmt_size = sizeof(struct ipc4_output_pin_format);
+			output_fmt_offset =
+				src_basecfg_ext->nb_input_pins * in_fmt_size +
+				bu->extension.r.src_queue * out_fmt_size;
+
+			memcpy_s(&out_fmt, sizeof(struct ipc4_output_pin_format),
+				 &src_basecfg_ext->pin_formats[output_fmt_offset],
+				 sizeof(struct ipc4_output_pin_format));
```

Comment:
seems we may not need use memcpy_s, can we directly use 
out_format = (struct ipc4_output_pin_format *)&src_basecfg_ext->pin_formats[output_fmt_offset];

Reply:
@btian1 is proposing to have `out_fmt` as a pointer and just point it into the array to read out the `obs`